File: stack/0880_decoded_string_at_index/0880_decoded_string_at_index.py

class Solution(object):
    def decodeAtIndex(self, S, K):
        i = 0
        for c in S:
            if c.isdigit():
                i *= int(c)
            else:
                i += 1
        
        for c in reversed(S):
            K %= i
            if K == 0 and c.isalpha():
                return c
            
            if c.isdigit():
                i /= int(c)
            else:
                i -= 1


S = "leet2code3"
K = 10
res = Solution().decodeAtIndex(S, K)
print(res)

File: stack/0880_decoded_string_at_index/0880_decoded_string_at_index.go
package main 

import (
	"fmt"
)

func decodeAtIndex(S string, K int) string {
	lenSi, i := 0, 0

	for ; lenSi < K; i++ {
		char := S[i]
		if isDigit(char) {
			lenSi *= int(char - '0')

		} else {
			lenSi++
		}
	}

	for {
		i--
		char := S[i]
		if isDigit(char) {
			lenSi /= int(char - '0')
			K %= lenSi
		} else {
			if K == 0 || K == lenSi {
				return string(char)
			}
		}
		lenSi--
	}
}

func isDigit(b byte) bool {
	return b <= '9'
}

func main() {
	S := "leet2code3"
	K := 10
	res := decodeAtIndex(S, K)
	fmt.Println(res)
}